activerecord/lib/active_record/relation/predicate_builder.rb

```diff
@@ -79,7 +81,20 @@ def expand_from_hash(attributes)
           if value.is_a?(Hash) && !table.has_column?(key)
             associated_predicate_builder(key).expand_from_hash(value)
           else
-            build(table.arel_attribute(key), value)
+            predicates = build(table.arel_attribute(key), value)
+
+            # Grouping NOT conditions that are of polymorphic associations
+            # in order to provide their proper inversion.
+            if of_inverted_clause &&
+              ((value.is_a?(AssociationQueryValue) &&
+                value.associated_table.polymorphic_association?) ||
+                  (value.is_a?(PolymorphicArrayValue) &&
+                    !predicates.is_a?(Arel::Nodes::Grouping)))
```

Comment:
I wonder if there's a better way to do this. It feels like a ton of work and is difficult to read.